universal-application-tool-0.0.1/app/repository/ApplicationRepository.java

```diff
@@ -5,62 +5,169 @@
-          application.save();
-          return application;
-        });
+          return submitApplicationInternal(applicant, program);
+        },
+        executionContext.current());
   }
 
   public CompletionStage<Optional<Application>> submitApplication(
       long applicantId, long programId) {
+    return this.perform(
+        applicantId,
+        programId,
+        (ApplicationArguments appArgs) ->
+            submitApplicationInternal(appArgs.applicant, appArgs.program));
+  }
+
+  private Application submitApplicationInternal(Applicant applicant, Program program) {
+    ebeanServer.beginTransaction();
+    try {
+      List<Application> oldApplications =
+          ebeanServer
+              .createQuery(Application.class)
+              .where()
+              .eq("applicant.id", applicant.id)
+              .eq("program.name", program.getProgramDefinition().name())
+              .findList();
+      for (Application application : oldApplications) {
+        // Delete any in-progress drafts, and mark obsolete any old applications.
+        if (application.getLifecycleStage().equals(LifecycleStage.DRAFT)) {
+          application.setLifecycleStage(LifecycleStage.DELETED);
+        } else {
+          application.setLifecycleStage(LifecycleStage.OBSOLETE);
+        }
+        application.save();
+      }
+      Application application = new Application(applicant, program, LifecycleStage.ACTIVE);
+      application.save();
+      ebeanServer.commitTransaction();
+      return application;
+    } finally {
+      ebeanServer.endTransaction();
+    }
+  }
+
+  private CompletionStage<Optional<Application>> perform(
+      long applicantId, long programId, Function<ApplicationArguments, Application> fn) {
     CompletionStage<Optional<Applicant>> applicantDb =
         applicantRepository.lookupApplicant(applicantId);
     CompletionStage<Optional<Program>> programDb = programRepository.lookupProgram(programId);
-    return applicantDb.thenCombineAsync(
-        programDb,
-        (applicantMaybe, programMaybe) -> {
-          if (applicantMaybe.isEmpty() || programMaybe.isEmpty()) {
-            return Optional.empty();
-          }
-          Applicant applicant = applicantMaybe.get();
-          Program program = programMaybe.get();
-          Application application = new Application(applicant, program, clock.instant());
-          application.save();
-          return Optional.of(application);
-        });
+    return applicantDb
+        .thenCombineAsync(
+            programDb,
+            (applicantMaybe, programMaybe) -> {
+              if (applicantMaybe.isEmpty()) {
+                throw new RuntimeException(new ApplicantNotFoundException(applicantId));
+              }
+              if (programMaybe.isEmpty()) {
+                throw new RuntimeException(new ProgramNotFoundException(programId));
+              }
+              return new ApplicationArguments(programMaybe.get(), applicantMaybe.get());
```

Comment:
Suggestion (non-blocking, optional): instead of making an inner class for this could do
```suggestion
              return Pair.of(programMaybe.get(), applicantMaybe.get());
```

Reply:
Gonna skip this one I think - Guava Pair.of() is not in the external Guava package, and is actually recommended against by the Guava folks, so we'd need to import someone else's Pair library.  When I tried that, I think I found that java type inference can't figure that out - it doesn't compile for some reason, at least.